src/monitor.ts

```diff
@@ -0,0 +1,425 @@
+ * a writeable jsonline file (.jl) as configuration. Each line should define a
+ * json object corresponding to the interface `IMonitoredNodeConfig`. The
+ * monitor process may modify the file and add more objects to it.
+ */
+
+import * as ChildProcess from 'child_process';
+import * as fs from 'fs-extra';
+import * as http from 'http';
+
+import * as bodyParser from 'body-parser';
+import * as program from 'commander';
+import * as express from 'express';
+import * as readline from 'readline';
+import * as rp from 'request-promise';
+import * as winston from 'winston';
+
+interface IMonitoredNodeConfig {
+  /**
+   * The rpc address of the monitored node. In IP:port format.
+   */
+  address: string;
+
+  /**
+   * A command that can be used to start/restart the node.  If `undefined`, the
+   * the monitor will not try to restart the node.
+   */
+  reviveCmd?: string;
+
+  /**
+   * The arguments that should be passed to the cmd above, seperated by space.
+   */
+  reviveArgs?: string;
+}
+
+/**
+ * Takes any object and checks if it conforms to the config interface.
+ */
+function validateMonitoredNodeConfig(data: any): IMonitoredNodeConfig {
+  if (!data.hasOwnProperty('address')) {
+    throw new Error('config does not have address field');
+  }
+
+  if (typeof(data.address) !== 'string') {
+    throw new Error('address is not a string');
+  }
+
+  if (data.hasOwnProperty('reviveCmd') &&
+      typeof(data.reviveCmd) !== 'string') {
+    throw new Error('invalid reviveCmd field');
+  }
+
+  if (data.hasOwnProperty('reviveArgs') &&
+      typeof(data.reviveArgs) !== 'string') {
+    throw new Error('invalid reviveArgs field');
+  }
+
+  return data;
+}
+
+/**
+ * An internal interface representing the state we keep track of for each
+ * network node.
+ */
+export interface IMonitoredNodeStatus {
+  address: string;
+  alive: boolean;
+  lastUpdate: Date;
+  lastResponseId: number;
+  failureCount: number;
+  networkId?: string;
+  reviveCmd?: string;
+  reviveArgs?: string[];
+}
+
+/**
+ * An internal interface representing the response of the ping command to the
+ * network node.
+ */
+interface INetVersionResponse {
+  id: number;
+  result: string;
+}
+
+/**
+ * The actual monitor implementation. It contains an express application and
+ * last known node status.
+ */
+export class Monitor {
+  /**
+   * The interval at which the monitor pings the instances. If the last update
+   * time was longer than this interval, the instance is considered dead.
+   */
+  private heartbeatInterval: number;
+
+  /**
+   * Number of failures the monitor should tolerate before considering a node as
+   * dead.
+   */
+  private failureTolerance: number;
+
+  // Path to the configuration file.
+  private configPath: string;
+
+  // The express app
+  private app: express.Application;
+
+  // The http server
+  private server?: http.Server;
+
+  // The current state of the nodes
+  private nodesStatus: IMonitoredNodeStatus[];
```

Comment:
nodeStatuses?

Reply:
also maybe we can still use /** stuff */ for one liner so if we auto-generate docs in the future they can be picked up.